File: God Zuo/Python/class052 单调栈/03.py
from typing import List

class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        MOD = 1000000007
        n = len(arr)
        ans = 0
        stack = []
        
        # 遍历阶段
        for i in range(n):
            # 注意：这里必须是 >=，这样才能在处理相同元素时，
            # 保证一侧严格小于，另一侧小于等于，避免重复计算相同范围的子数组
            while stack and arr[stack[-1]] >= arr[i]:
                cur = stack.pop()
                # 弹出时计算 cur 位置作为最小值的辐射范围
                left = stack[-1] if stack else -1
                
                # (cur - left) 是向左辐射的长度，(i - cur) 是向右辐射的长度
                # 组合数原理：左边的情况数 * 右边的情况数
                ans = (ans + (cur - left) * (i - cur) * arr[cur]) % MOD
                
            stack.append(i)  
            
        # 清算阶段
        while stack:
            cur = stack.pop()
            left = stack[-1] if stack else -1
            # 清算阶段意味着右侧没有遇到比它更小的值，所以右边界直接伸展到数组末尾 n
            ans = (ans + (cur - left) * (n - cur) * arr[cur]) % MOD
            
        return ans  

File: Testing/Mati/正赛/国赛模拟/03.cpp
#include<bits/stdc++.h>
using namespace std;

int n;
string s;
int MOD = 998244353;
int dp[5005][5005];

int f(int l,int r){
    if(l==r)
        return 1;
    if(dp[l][r]!=-1)
        return dp[l][r];
    int ans;
    if(s[l]==s[r]){
        ans=(f(l+1,r)+f(l,r-1))
    }
}

int main(){
    cin >> n;
    cin >> s;
    for (int i = 0; i < n; ++i){
        dp[i][i] = 1;

    }
}